File: problems/回溯/47.全排列II.py
#!/usr/bin/env python3
"""
47. 全排列 II (Permutations II)
题目: 给定一个可包含重复数字的序列，返回所有不重复的全排列。
算法: 回溯 + 排序 + 剪枝
时间复杂度: O(n! * n)
空间复杂度: O(n)
"""

def permuteUnique(nums):
    result = []
    path = []
    used = [False] * len(nums)
    nums.sort()  # 排序用于去重
    
    def backtrack():
        if len(path) == len(nums):
            result.append(path[:])
            return
        for i in range(len(nums)):
            if used[i]:
                continue
            # 剪枝：相同元素不能重复使用
            if i > 0 and nums[i] == nums[i - 1] and not used[i - 1]:
                continue
            used[i] = True
            path.append(nums[i])
            backtrack()
            path.pop()
            used[i] = False
    
    backtrack()
    return result

if __name__ == "__main__":
    # 测试用例1
    print(permuteUnique([1, 1, 2]))
    # 输出: [[1,1,2], [1,2,1], [2,1,1]]
    
    # 测试用例2
    print(permuteUnique([1, 2, 3]))
    # 输出: 6个排列
    
    # 测试用例3
    print(permuteUnique([2, 2, 1, 1]))
    # 输出: 12个排列

File: problems/回溯/47.全排列II.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

void backtrack(vector<vector<int>>& result, vector<int>& path,
               vector<int>& nums, vector<bool>& used) {
    if (path.size() == nums.size()) {
        result.push_back(path);
        return;
    }
    for (int i = 0; i < nums.size(); i++) {
        if (used[i]) continue;
        // 剪枝：相同元素跳过
        if (i > 0 && nums[i] == nums[i - 1] && !used[i - 1]) continue;
        
        used[i] = true;
        path.push_back(nums[i]);
        backtrack(result, path, nums, used);
        path.pop_back();
        used[i] = false;
    }
}

vector<vector<int>> permuteUnique(vector<int>& nums) {
    vector<vector<int>> result;
    vector<int> path;
    vector<bool> used(nums.size(), false);
    sort(nums.begin(), nums.end());  // 排序用于去重
    backtrack(result, path, nums, used);
    return result;
}

int main() {
    // 测试用例1
    vector<int> nums1 = {1, 1, 2};
    vector<vector<int>> res1 = permuteUnique(nums1);
    cout << "Test 1: [1,1,2] - " << res1.size() << " unique permutations" << endl;
    for (auto& perm : res1) {
        cout << "[";
        for (int i = 0; i < perm.size(); i++) {
            cout << perm[i];
            if (i < perm.size() - 1) cout << ",";
        }
        cout << "] ";
    }
    cout << endl;
    
    // 测试用例2
    vector<int> nums2 = {1, 2, 3};
    vector<vector<int>> res2 = permuteUnique(nums2);
    cout << "Test 2: [1,2,3] - " << res2.size() << " permutations" << endl;
    
    // 测试用例3
    vector<int> nums3 = {2, 2, 1, 1};
    vector<vector<int>> res3 = permuteUnique(nums3);
    cout << "Test 3: [2,2,1,1] - " << res3.size() << " unique permutations" << endl;
    
    return 0;
}
